Feature/add metastore datasource

guillesd opened the pull request:
Addresses #1651

This PR would enable users to retrieve existing metastores, which will allow them to get the needed ID of a Metastore by name. This ID can then be used to assign a metastore to a workspace, without the need of hardcoding an ID for different workspace deployments.

nkvuong commented:
@guillesd we should wait until the account API for Unity Catalog is available before implementing this data source.

for this PR, you also need to add the mapping to [provider.go]https://github.com/databricks/terraform-provider-databricks/tree/master/provider/provider.go

guillesd commented:
Hi @nkvuong! As discussed in the issue #1651, every unity endpoint uses the Workspace API, so i don't see why we shouldn't make this data source available. Even the `databricks_metastore` resource uses the Workspace API

guillesd commented:
@nkvuong I added the mapping to provider.go

alexott commented on docs/data-sources/metastores.md:
```diff
@@ -0,0 +1,33 @@
+
+This data source exports the following attributes:
+
+* `metastores` - list of [databricks_metastore](../resources/share.md)
```
it's pointing to the wrong file

alexott commented on catalog/data_metastores.go:
```diff
@@ -0,0 +1,37 @@
+			return err
+		}
+		for _, v := range metastores.Metastores {
+			data.Name = append(data.Name, v.name)
+			data.StorageRoot = append(data.StorageRoot, v.storage_root)
+			data.DefaultDacID = append(data.DefaultDacID, v.default_data_access_config_id)
+			data.Owner = append(data.Owner, v.owner)
+			data.MetastoreID = append(data.MetastoreID, v.metastore_id)
+			data.Region = append(data.Region, v.region)
+			data.Cloud = append(data.Cloud, v.cloud)
+			data.GlobalMetastoreId = append(data.GlobalMetastoreId, v.global_metastore_id)
+			data.CreatedAt = append(data.CreatedAt, v.created_at)
+			data.CreatedBy = append(data.CreatedBy, v.created_by)
+			data.UpdatedAt = append(data.UpdatedAt, v.updated_at)
+			data.UpdatedBy = append(data.UpdatedBy, v.updated_by)
+			data.DeltaSharingScope = append(data.DeltaSharingScope, v.delta_sharing_scope)
+			data.DeltaSharingRecipientTokenLifetimeInSeconds = append(data.DeltaSharingRecipientTokenLifetimeInSeconds, v.delta_sharing_recipient_token_lifetime_in_seconds)
+			data.DeltaSharingOrganizationName = (data.DeltaSharingOrganizationName, v.delta_sharing_organization_name)
```
How the data is look like in use? I don't see where do you expose the `metastores` attribute...

alexott commented on catalog/data_metastores_test.go:
```diff
@@ -0,0 +1,53 @@
+		Read:        true,
+		NonWritable: true,
+		ID:          "_",
+	}.ApplyNoError(t)
```
Also, add a check for the data returned

nfx commented:
SDK is merged in #1848 , please update code accordingly

guillesd commented:
@alexott I fixed:
- test/linting errors (sorry for this, my bad)
- wrong reference in docs
I also added:
- A specific data test

@nfx not sure what you mean by "SDK is merged update code accordingly"

nfx commented:
@guillesd it means that you have to use https://pkg.go.dev/github.com/databricks/databricks-sdk-go@v0.3.2/service/unitycatalog#MetastoresAPI 

e.g.

```go
w, err := c.WorkspaceClient()
..
w.Metastores....
```

guillesd commented:
ok @nfx, so if I understand correctly then all the API functions with this type of structure:
```go
func (a MetastoresAPI) listMetastores() (mis Metastores, err error) {
	err = a.client.Get(a.context, "/unity-catalog/metastores", nil, &mis)
	return
}
```
will be replaced with their SDK counterpart?

nfx commented:
> will be replaced with their SDK counterpart?

@guillesd that's correct

nfx commented:
@guillesd all tests must pass

guillesd commented on catalog/data_metastores.go:
```diff
@@ -0,0 +1,25 @@
+
+func DataSourceMetastores() *schema.Resource {
+	type MetastoresData struct {
+		Metastores []unitycatalog.MetastoreInfo `json:"metastores,omitempty" tf:"computed"`
```
I've seen indeed that this is something that is done with other data sources... but for me this is not really nice for the user, since they would need two resource blocks to get the information of my metastores (first the `data_metastores`, then the `data_metastore` passing the specific ID of the metastore).

guillesd commented:
@nfx yes, all test should pass. But the specific tests that are failing have nothing to do with my addition, there's just some test that fail for weird reasons:
```
=== FAIL: provider TestConfig_ConfigFile (0.00s)
2023/03/17 14:25:22 [INFO] Explicit and implicit attributes: 
2023/03/17 14:25:22 [TRACE] Loading config via environment
2023/03/17 14:25:22 [TRACE] Loading config via config-file
2023/03/17 14:25:22 [DEBUG] Loading DEFAULT profile from /Users/guillermosanchez/.databrickscfg
2023/03/17 14:25:22 [TRACE] Attempting to configure auth: pat
```
This is trying to connect to my default databricks CLI profile, but why?

guillesd commented:
@nfx anything blocking this?

nkvuong commented:
@guillesd we now have account-level API to list metastores in the SDK, so it would make sense to use that instead
https://github.com/databricks/databricks-sdk-go/blob/main/service/unitycatalog/api.go#L189

guillesd commented:
@nkvuong this is not released yet for what I see, I should wait for that I guess!

nfx commented:
@guillesd it is released :)

alexott commented:
@guillesd would you able to finish this PR?

guillesd commented:
I'm taking a look at it now @alexott. Trying to find an example of another resource that uses the account API and the SDK

guillesd commented on catalog/data_metastores.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+func DataSourceMetastores() *schema.Resource {
+	return common.WorkspaceData(func(ctx context.Context, data *MetastoresData, w *databricks.WorkspaceClient) error {
```
So @alexott here I'm switching to `*databricks.AccountClient` but I'm having a weird issue now. Can I use `common.WorkspaceData` with the `AccountClient` context?

tanmay-db replied:
Is this outdated? What was the context / discussion behind this

tanmay-db commented:
Hi @guillesd, thanks for creating a PR on this. If it's okay with you, I will add changes to your PR and address the comments / integration tests etc...

tanmay-db commented on internal/acceptance/data_metastores_test.go:
```diff
@@ -0,0 +1,16 @@
+	"testing"
+)
+
+func TestAccDataSourceMetastore(t *testing.T) {
+	workspaceLevel(t, step{
+		Template: `
+		data "databricks_metastores" "this" {}
+		
+		output "all_metastores" {
+			value = data.databricks_metastores.this
+		}`,
+	})
+}
```
Using output should be fine, used cluster test for reference: https://github.com/databricks/terraform-provider-databricks/blob/master/internal/acceptance/data_cluster_test.go

tanmay-db replied:
Tested locally, working fine

tanmay-db replied:
Also created a run on CI

tanmay-db commented on provider/provider.go:
```diff
@@ -62,6 +62,7 @@ func DatabricksProvider() *schema.Provider {
 			"databricks_instance_pool":           pools.DataSourceInstancePool(),
 			"databricks_jobs":                    jobs.DataSourceJobs(),
 			"databricks_job":                     jobs.DataSourceJob(),
+			"databricks_metastores":              catalog.DataSourceMetastores(),
```
should we have `databricks_metastores` or `databricks_metastore`?

tanmay-db replied:
most likely metastore`s` since there can be multiple of them but highlighting for visibility

alexott replied:
It should be `metastores` because it returns a list. A standard pattern usually is to have two data sources - one returning a list, and another searching for a specific resource and throwing an error if it's not found

tanmay-db replied:
I will add that in another PR as it will require docs and tests and we can merge this before that.

alexott commented on catalog/data_metastores_test.go:
```diff
@@ -0,0 +1,93 @@
+		Read:        true,
+		NonWritable: true,
+		ID:          "_",
+		HCL: `metastores {
+			name = "a"
+			storage_root = "abc"
+		}`,
```
why it's required?

alexott reviewed (approved):
code wise looks good, but really maybe we need to add some filtering capabilities? Like, search by substring in the name, etc.?

tanmay-db commented:
@alexott makes sense, I will add databricks_metastore resource as well 
@nkvuong thanks, will move it to be a account level resource.

tanmay-db commented:
https://github.com/databricks/terraform-provider-databricks/pull/2429/files

tanmay-db commented on catalog/data_metastores_test.go:
```diff
@@ -0,0 +1,92 @@
+		Fixtures: []qa.HTTPFixture{
+			{
+				Method:   "GET",
+				Resource: "/api/2.0/accounts/testaccount/metastores",
```
We use 2.0 https://github.com/databricks/databricks-sdk-go/blob/main/service/catalog/impl.go#L78

tanmay-db commented on catalog/data_metastores.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+func DataSourceMetastores() *schema.Resource {
+	return common.AccountData(func(ctx context.Context, data *MetastoresData, acc *databricks.AccountClient) error {
```
This is added in https://github.com/databricks/terraform-provider-databricks/pull/2429/files

tanmay-db commented on catalog/data_metastores.go:
```diff
@@ -0,0 +1,20 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func DataSourceMetastores() *schema.Resource {
+	return common.AccountData(func(ctx context.Context, data *MetastoresData, acc *databricks.AccountClient) error {
+		metastores, err := acc.Metastores.List(ctx)
+		if err != nil {
+			return err
+		}
+		data.Metastores = metastores.Metastores
+		return nil
+	})
```
cc: @mgyucht

tanmay-db commented on catalog/resource_metastore.go:
```diff
@@ -38,16 +38,15 @@ type MetastoreInfo struct {
 	StorageRoot string `json:"storage_root"`
 }
 
-// func (a MetastoresAPI) listMetastores() (mis []MetastoreInfo, err error) {
-// 	err = a.client.Get(a.context, "/unity-catalog/metastores", nil, &mis)
-// 	return
-// }
-
```
remove dead code, we already have crud

mgyucht reviewed (commented):
Let's double check with UC team how they feel about data sources implemented with accounts client only before going forward with this.

adamcain-db commented:
> Let's double check with UC team how they feel about data sources implemented with accounts client only before going forward with this.

On behalf of the UC team, we are quite in favor of the terraform provider using the account-level UC Metastore CRUD/assign APIs instead of the workspce-level UC Metastore APIs.  Thanks!

mgyucht commented on internal/acceptance/data_metastores_test.go:
```diff
@@ -0,0 +1,16 @@
+func TestUcAccDataSourceMetastore(t *testing.T) {
+	accountLevel(t, step{
+		Template: `
+		data "databricks_metastores" "this" {}
```
This is a reasonable integration test for the "no metastores" case. Can we add an integration test for the "one or more metastores case"?

nkvuong replied:
@mgyucht not sure how easy it is to test both scenarios in a single account, unless we add a filter option to this data source

tanmay-db replied:
Check has been added in the test

mgyucht reviewed (commented):
Almost good to go, but let's improve the integration test.

nkvuong commented:
Account-level APIs for UC now works without the additional header 🎉 

Acceptance tests passed

nkvuong reviewed (changes requested):
it would make more sense to return a list of name -> id mapping, otherwise the id on itself is not very useful

tanmay-db commented on catalog/data_metastores.go:
```diff
@@ -0,0 +1,32 @@
+			name := v.Name
+			_, duplicateName := data.Ids[name]
+			if duplicateName {
+				return fmt.Errorf("duplicate metastore name detected: %s", name)
```
cc: @nkvuong on this for visiblity
